add tutorial about sing Javalin as a simulator for HTTP-based APIs

lsoares opened the pull request.

tipsy commented:
Thanks @lsoares!

Did you see the `javalin-testtools` module? These are the tests for that module:

```kotlin
 @Test
    fun `get method works`() = TestUtil.test { server, client ->
        server.get("/hello") { it.result("Hello, World!") }
        val response = client.get("/hello")
        assertThat(response.code).isEqualTo(200)
        assertThat(response.body!!.string()).isEqualTo("Hello, World!")
    }

    @Test
    fun `post with json serialization works`() = TestUtil.test { server, client ->
        server.post("/hello") { it.result(it.bodyAsClass<MyKotlinClass>().field1) }
        val response = client.post("/hello", MyKotlinClass("v1", "v2"))
        assertThat(response.body?.string()).isEqualTo("v1")
    }

    @Test
    fun `request method works`() = TestUtil.test { server, client ->
        server.post("/form") { it.result(it.formParam("username")!!) }
        val response = client.request("/form") {
            it.post(FormBody.Builder().add("username", "test").build())
        }
        assertThat(response.body!!.string()).isEqualTo("test")
    }

    @Test
    fun `custom javalin works`() {
        val app = Javalin.create()
                .get("/hello") { it.result("Hello, World!") }
        TestUtil.test(app) { server, client ->
            assertThat(client.get("/hello").body?.string()).isEqualTo("Hello, World!")
        }
    }
```

Maybe you could use it, or expand on your approach to use a method similar to ```fun `request method works`() = TestUtil.test { server, client -``` ? I think it might result in cleaner code :)

lsoares commented:
hi @tipsy 
are you referring to [this](https://mvnrepository.com/artifact/io.javalin/javalin-testtools)? it looks to be alpha but I can give it a try.

I tried converting the tests and they look like this:
```kotlin
@Test
fun `gets a user profile by id v2`() = TestUtil.test { server, _ ->
    server.get("profile/abc") {
        it.json(mapOf("id" to "abc", "email" to "x123@gmail.com"))
    }
    val profileGateway = ProfileGateway(apiUrl = "http://localhost:${server.port()}")

    val result = profileGateway.fetchProfile("abc")

    assertEquals(Profile(id = "abc", email = "x123@gmail.com".toEmail()), result)
}
```

```kotlin
@Test
fun `posts a user profile v2`() = TestUtil.test { server, _ ->
    var postedBody: String? = null
    var contentType: String? = null
    server.post("profile") {
        postedBody = it.body()
        contentType = it.contentType()
        it.status(201)
    }
    val profileGateway = ProfileGateway(apiUrl = "http://localhost:${server.port()}")

    profileGateway.saveProfile(Profile(id = "abc", email = "x123@gmail.com".toEmail()))

    JSONAssert.assertEquals(
        """ { "id": "abc", "email": "x123@gmail.com"}  """,
        postedBody, true
    )
    assertEquals("application/json", contentType)
}
```

lsoares commented:
my review:

👍 no need to declare the server globally
👍 no need to start and stop the server (in` @AfterEach)`
👎 one more library for tests - [still in alpha in Maven Repository](https://mvnrepository.com/artifact/io.javalin/javalin-testtools)
👎 a little bit of magic that one has to understand (but worth it for the removal of boilerplate code)
👎 `client` is not needed and has to be declared as `_`

what do you think? I'm skeptical/divided.

tipsy commented:
I agree with your pro/con list - it's probably better if you use the method for inspiration than if you actually use it outright. You could create your own `TestUtil.test` so you get:

```kotlin
@Test
fun `gets a user profile by id v2`() = TestUtil.test { mockGateway, gatewayClient ->
    mockGateway.get("profile/abc") {
        it.json(mapOf("id" to "abc", "email" to "x123@gmail.com"))
    }
    val result = gatewayClient.fetchProfile("abc")
    assertEquals(Profile(id = "abc", email = "x123@gmail.com".toEmail()), result)
}

@Test
fun `posts a user profile v2`() = TestUtil.test { mockGateway, gatewayClient ->
    var contentType: String? = null
    server.post("profile") {
        postedBody = it.body()
        contentType = it.contentType()
        it.status(201)
    }

    gatewayClient.saveProfile(Profile(id = "abc", email = "x123@gmail.com".toEmail()))

    JSONAssert.assertEquals(
        """ { "id": "abc", "email": "x123@gmail.com"}  """,
        postedBody, true
    )
    assertEquals("application/json", contentType)
}
```

Then you could include a small comment about the `testtools` becoming available in Javalin 4? :)

Overall I *strongly* recommend using a method similar to `TestUtil.test`, almost all of Javalin's own tests use this method, it's very much worth if if you're going to be writing more than a handful of tests. The method itself can be fairly simple: 

```kotlin
@JvmStatic
fun test(app: Javalin, test: ThrowingBiConsumer<Javalin, HttpClient>) {
    app.start(0)
    val http = HttpClient(app)
    test.accept(app, http) // this is where the user's test happens
    app.stop()
}
```

lsoares commented:
the util looks idea good! but... it still depends on `ThrowingBiConsumer` and it does not target my gateway client.
given that, I converted it. this is how it looks:

```kotlin
@Test
fun `gets a user profile by id v2`() = test { server, gatewayClient ->
    server.get("profile/abc") {
        it.json(mapOf("id" to "abc", "email" to "x123@gmail.com"))
    }

    val result = gatewayClient.fetchProfile("abc")

    assertEquals(Profile(id = "abc", email = "x123@gmail.com".toEmail()), result)
}

private fun test(block: (Javalin, ProfileGateway) -> Unit) {
    val server = Javalin.create().start(0)
    val gatewayClient = ProfileGateway(apiUrl = "http://localhost:${server.port()}")
    block(server, gatewayClient)
    server.stop()
}
```

👎 it hides the creation of the test subject
👍 it hides boilerplate code

As you said, it gets progressively better if we wanted to test lots of variations of the gateway client.
what if I add a section to the article explaining that util method and hinting about `testtools`?

tipsy commented:
> 👎 it hides the creation of the test subject

This is a good thing if that creation is the same for every test.

> I converted it. this is how it looks:

Looks good, but be aware that it will only work from Kotlin (Java users would have to use try/catch)

> what if I add a section to the article explaining that util method and hinting about testtools?

That would be nice :)

tipsy commented:
I just did some cleanup to the `testtools` module. If you're interested in making it work from Java too, you can make an interface like:

```
fun interface TestCase {
    @Throws(Exception::class)
    fun accept(server: Javalin, client: ApiGatewayClient)
}
```

lsoares commented:
just added a new commit with that approach.

- I added that the approach is inspired `javalin-testtools` (btw do you have a link?); not sure what else to say about it since it has another goal - it's not focused on that test subject.
- I didn't add the Java part as the whole tutorial is Kotlin (it's in the metadata) and it would feel weird to just talk about it there in the end
- feel free to make a commit to improve the tutorial, as English is not my native language

thanks for the discussion about the improvements.

tipsy commented:
This looks great @lsoares, I like the way to chose to introduce the utility method as something that's useful for bigger test suites. I'll merge this now, and make a few small adjustments to formatting and language. Nice work :)

tipsy commented:
It's live! https://javalin.io/tutorials/using-javalin-as-http-simulator

lsoares commented:
Awesome!
